tplinkrouterc6u/client.py

```diff
@@ -151,22 +167,64 @@ def _calc_cpu_usage(data: dict) -> float | None:
 
         for item in data.get('access_devices_wireless_host', []):
             type = Wifi.WIFI_2G if '2.4G' == item['wire_type'] else Wifi.WIFI_5G
-            status.devices.append(Device(type, item['macaddr'], item['ipaddr'], item['hostname']))
+            status.devices.append(Device(type, macaddress.EUI48(item['macaddr']), ipaddress.IPv4Address(item['ipaddr']), item['hostname']))
 
         for item in data.get('access_devices_wireless_guest', []):
             type = Wifi.WIFI_GUEST_2G if '2.4G' == item['wire_type'] else Wifi.WIFI_GUEST_5G
-            status.devices.append(Device(type, item['macaddr'], item['ipaddr'], item['hostname']))
+            status.devices.append(Device(type, macaddress.EUI48(item['macaddr']), ipaddress.IPv4Address(item['ipaddr']), item['hostname']))
 
         return status
-
+    
+    def _get_ipv4_status(self) -> IPv4Status:
+        ipv4_status = IPv4Status
+        data = self._get_data('admin/network?form=status_ipv4', 'operation=read')
+        ipv4_status.wan_macaddr = macaddress.EUI48(data['wan_macaddr'])
+        ipv4_status.wan_ipv4_ipaddr = ipaddress.IPv4Address(data['wan_ipv4_ipaddr'])
+        ipv4_status.wan_ipv4_gateway = ipaddress.IPv4Address(data['wan_ipv4_gateway'])
+        ipv4_status.wan_ipv4_conntype = data['wan_ipv4_conntype']
+        ipv4_status.wan_ipv4_netmask = ipaddress.IPv4Address(data['wan_ipv4_netmask'])
+        ipv4_status.wan_ipv4_pridns = ipaddress.IPv4Address(data['wan_ipv4_pridns'])
+        ipv4_status.wan_ipv4_snddns = ipaddress.IPv4Address(data['wan_ipv4_snddns'])
+        ipv4_status.lan_macaddr = macaddress.EUI48(data['lan_macaddr'])
+        ipv4_status.lan_ipv4_ipaddr = ipaddress.IPv4Address(data['lan_ipv4_ipaddr'])
+        ipv4_status.lan_ipv4_dhcp_enable = self._str2bool(data['lan_ipv4_dhcp_enable'])
+        ipv4_status.lan_ipv4_netmask = ipaddress.IPv4Address(data['lan_ipv4_netmask'])
+        ipv4_status.remote = self._str2bool(data['remote'])
+
+        return ipv4_status
+
+    def _get_ipv4_reservations(self) -> [IPv4Reservation]:
+        ipv4_reservations = []
+        data = self._get_data('admin/dhcps?form=reservation', 'operation=load')
+
+        for item in data:
+            ipv4_reservations.append(IPv4Reservation(macaddress.EUI48(item['mac']), ipaddress.IPv4Address(item['ip']), item['comment'], self._str2bool(item['enable'])))
```

Comment:
Is there a way to implement sort method in IPv4Reservation to sort by ip without using macaddress and ipaddress? I made this client for Home Assistant component https://github.com/AlexandrErohin/home-assistant-tplink-router and Home Assistant work with ip and macaddress as string, so with this additions it would need to cast every time to str

Reply:
I kept the original interface properties to the dataclasses and added @property decorators to return the macaddress/ipaddress datatypes. I also made all the xxxStatus dataclasses with __init__() to return objects rather than use class attributes (this keeps everything following the same pattern). I also added a test.py utility to validate my changes fit your HA component use. The utility also runs through a list of queries in queries.txt to call every cgi-bin form and log the dictionary result to a corresponding file in the logs folder. I captured every click in the UI using the chrome developer tools and saved to a .har file. Then used that to find all the "/admin/foo?form=bar" URIs and call them using the client query(query, operation) API. Hope you like the additions.